### PR TITLE
Add a channel get/put hoisting pass

### PR DESCRIPTION
**Pull request by fifield**
Some experiments hoisting `air.channel.get/put` out of loops by adding an extra dimension to the op.

In addition this adds some methods to the `air::ChannelInterface` interface

### CODE REVIEW

File: mlir/lib/Transform/AIRMiscPasses.cpp [hunk 1]
```diff
@@ -879,6 +881,119 @@ void AIRLowerHerdParallelPass::runOnOperation() {
...
+
+    // channel op must be the only op in the loop
+    if (forOp.getBody()->getOperations().size() != 2)
+      return failure();
+
+    // scf.for must have constant bounds and step
+    auto lb = forOp.getLowerBound().getDefiningOp<arith::ConstantIndexOp>();
+    auto ub = forOp.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+    auto step = forOp.getStep().getDefiningOp<arith::ConstantIndexOp>();
+    if (!lb || !ub || !step)
```
Comment (keryell) [inline mlir/lib/Transform/AIRMiscPasses.cpp:909]:
```suggestion
    if (!(lb && ub && step))
```
might be easier to understand the precondition you expect.

File: mlir/lib/Transform/AIRMiscPasses.cpp [hunk 2]
```diff
@@ -879,6 +881,119 @@ void AIRLowerHerdParallelPass::runOnOperation() {
...
+          offsets, sizes, strides);
+    else
+      rewriter.create<air::ChannelGetOp>(
+          loc, tys, deps, channel, chanOp.getIndices(), chanOp.getMemref(),
+          offsets, sizes, strides);
+    rewriter.eraseOp(op);
+    rewriter.eraseOp(forOp);
+    return success();
+  }
+};
```
Comment (keryell) [inline mlir/lib/Transform/AIRMiscPasses.cpp:974]: Ouch, 85 lines of mostly uncommented code... Feel free to help some eventual readers. :-)

File: mlir/lib/Transform/AIRMiscPasses.cpp (hunk 1 above)
Comment (fifield) [reply mlir/lib/Transform/AIRMiscPasses.cpp:909]: I prefer the one I can parse left-to-right without a mental stack.

Comment (keryell) [reply mlir/lib/Transform/AIRMiscPasses.cpp:909]: I did not think you were a Turing machine! ;-)

**Pull request merged**